query_optimizer/ExecutionGenerator.cpp

```diff
@@ -332,6 +336,55 @@ void ExecutionGenerator::createTemporaryCatalogRelation(
   insert_destination_proto->set_relation_id(output_rel_id);
 }
 
+void ExecutionGenerator::createPartitionedTemporaryCatalogRelation(
+    const P::PhysicalPtr &physical,
+    const CatalogRelation **catalog_relation_output,
+    S::InsertDestination *insert_destination_proto,
+    const attribute_id part_attr_id,
+    const std::size_t num_partitions) {
+  std::unique_ptr<CatalogRelation> catalog_relation(
+      new CatalogRelation(optimizer_context_->catalog_database(),
+                          getNewRelationName(),
+                          -1 /* id */,
+                          true /* is_temporary*/));
+  attribute_id aid = 0;
+  for (const E::NamedExpressionPtr &project_expression :
+       physical->getOutputAttributes()) {
+    // The attribute name is simply set to the attribute id to make it distinct.
+    std::unique_ptr<CatalogAttribute> catalog_attribute(
+        new CatalogAttribute(catalog_relation.get(),
+                             std::to_string(aid),
+                             project_expression->getValueType(),
+                             aid,
+                             project_expression->attribute_alias()));
+    attribute_substitution_map_[project_expression->id()] =
+        catalog_attribute.get();
+    catalog_relation->addAttribute(catalog_attribute.release());
+    ++aid;
+  }
+
+  PartitionSchemeHeader *part_scheme_header = new HashPartitionSchemeHeader(num_partitions, part_attr_id);
```

Comment:
I was wondering if we have a particular reason to use `HashPartition` by default. As I explain before, we may need to derive the partition info from the input relation.